### PR TITLE
[RPC] Only return hex field once in getrawtransaction

### PR DESCRIPTION
**Pull request by achow101**
The hex is already returned in `TxToUniv()`, no need to give it out a second time in getrawtransaction itself.

### CODE REVIEW

**Comment by sdaftuar** [comment]
ACK

**Comment by achow101** [comment]
@MarcoFalke Hmm. That might be a problem since `RPCSerializationFlags` are not passed to `EncodeHexTx` in `TxToUniv`.

**Comment by sipa** [comment]
> That might be a problem since `RPCSerializationFlags` are not passed to `EncodeHexTx` in `TxToUniv`.

That seems to be a bug on its own.

**Comment by achow101** [comment]
Another thing is that not everything that uses `TxToUniv` needs to have hex output, e.g. `decoderawtransaction`

**Comment by achow101** [comment]
I added a commit which passes the `RPCSerializationFlags` to `TxToUniv` and a boolean for whether to give `hex` output or not.

File: src/core_io.h [hunk 1]
```diff
@@ -31,6 +31,6 @@ UniValue ValueFromAmount(const CAmount& amount);
 std::string FormatScript(const CScript& script);
 std::string EncodeHexTx(const CTransaction& tx, const int serializeFlags = 0);
 void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool fIncludeHex);
-void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry);
+void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, const int serializeFlags = 0, bool include_hex = true);
```
Comment (sipa) [inline src/core_io.h]: Marking a by-value parameter as `const` is redundant. Also, style nit on `serializeFlags`.

**Review by sipa — COMMENTED** [review]
utACK

File: src/core_io.h (hunk 1 above)
Comment (achow101) [inline src/core_io.h]: Fixed. It was copy and pasted from two lines above in `EncodeHexTx` :)

**Comment by MarcoFalke** [comment]
> That seems to be a bug on its own.

I don't think so?

According to the doc, `-rpcserialversion` sets the serialization of raw transaction or block hex returned in        **non-verbose** mode.

**Review by promag — COMMENTED** [review]
utACK 9381d86.

There are only 2 more calls to `TxToUniv` so I wonder if it pays to have default arguments.

**Comment by achow101** [comment]
@MarcoFalke I interpret that as talking about when verbose mode means that you wouldn't get the hex back, e.g. in `getblock`. If we are returning hex, then I think we should be consistent and all hex should follow whatever the `RPCSerializationFlags` are.

**Review by gmaxwell — APPROVED** [review]
ACK.  This needs a 0.15 tag: it's a 0.15 regression and it results in invalid JSON.

File: src/core_io.h [hunk 2]
```diff
@@ -31,6 +31,6 @@ UniValue ValueFromAmount(const CAmount& amount);
 std::string FormatScript(const CScript& script);
 std::string EncodeHexTx(const CTransaction& tx, const int serializeFlags = 0);
 void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool fIncludeHex);
-void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry);
```
Comment (laanwj) [inline src/core_io.h]: Don't care about the order, though I'd prefer not using default arguments. Easy enough to just add these on all call sites, there aren't many.

**Comment by jnewbery** [comment]
Strictly speaking, I don't think this is invalid JSON, although client behaviour is undefined if object names are not unique (https://tools.ietf.org/html/rfc7159#section-4). But yes, this is a regression in 0.15 and should be treated as a bug fix.

Perhaps we should add JSON object checking to rpcserver.cpp (in `JSONRPCExecOne()`) and/or authproxy.py (in `__call__()`) to verify the validity of our returned JSON.

**Comment by jnewbery** [comment]
Concept ACK

> Another thing is that not everything that uses TxToUniv needs to have hex output, e.g. decoderawtransaction
I added a commit which passes ... a boolean for whether to give hex output or not.

I don't think you should complicate the interface to `TxToUniv` to specify which fields to return. Either filter them out at the caller, or just return the hex in `decoderawtransaction`. Is there any harm in returning the hex in that RPC?

**Comment by achow101** [comment]
Swapped the arguments.

>  Either filter them out at the caller

I looked at that but I couldn't find anything in Univalue for removing things from it.

> or just return the hex in `decoderawtransaction`. Is there any harm in returning the hex in that RPC?

Since you are already providing it the hex, it's not useful to be giving the hex back out to the user.

**Comment by jtimon** [comment]
utACK e029c6e

> There are only 2 more calls to TxToUniv so I wonder if it pays to have default arguments.

+1

**Comment by promag** [comment]
Wild option 🙄:
```
void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry);
void TxToUnivWithHex(const CTransaction& tx, const uint256& hashBlock, int serialize_flags, UniValue& entry);
```

**Comment by laanwj** [comment]
> Since you are already providing it the hex, it's not useful to be giving the hex back out to the user.

I was also really surprised to see the hex being echoed back in `decoderawtransaction` output. That's just a waste of bandwidth especially for large transactions.

>  Either filter them out at the caller

No, don't do this. univalue objects aren't meant to be written to/processed after initial generation. The library is not optimized for that. If you need any kind of editing use an intermedaite data structure.

**Pull request merged**